### DISCUSSION TITLE
Question: full text search

### DISCUSSION THREAD

**Discussion by amityahav** [asciinema/discussions#237] [Q&A]
Hey, we use asciinema in production to display ssh  session recordings of users.
One feature that we want to implement is text search over those sessions and highlighting the timestamps where the searched keywords/phrases occur. Implementing this Using the raw ascicast file will be quite hard since it captures keystrokes which are hard to combine to a word, so my second thought (inefficient though) is dumping the entire state of the virtual terminal in each timestamp and index the keywords in some efficient way. So my questions are:
1. Do you think the second approach is feasible?
2. Does the dump function is public and can be exported to the asciiplayer? 
3. Does the dump function returns all the text that is visible in the terminal at a current state? (without scrolling)

Thanks, Amit

**Comment by amityahav** [comment]
@ku1ik  hey, just saw ur response on that subject on another issue. would appreciate your response here as well!

**Comment by ku1ik** [comment]
Hey Amit,

I believe the only reliable way is indeed to convert the asciicast to text representation. So, in fact if you want per-frame search you would need to take a text snapshot of terminal text contents on every frame. 

`dump` method on `Vt` doesn't dump just text but it's a full dump of terminal emulator _state_. So in addition to text, it also includes control sequences to turn Vt into the exact same state when passed to `vt.feed_str`. In other words, `dump` is kind of a "reverse of `feed_str` on initial Vt". Given that, this is not going to help for you use case, even if `dump` was exported as a public method.

What you're looking for is probably `vt.lines()`, which returns current screen contents. For each output event from asciicast file you would `vt.feed_str(...)` followed by `vt.lines()`. 

Note that `avt` is still pre-1.0 so the API for this may change (although if it does then probably only slightly, as getting current text from avt is what the player needs).

Your use case is pretty cool, but something like that would require significant amount of work around the player/avt, so not something I plan to add anytime soon.

Unless... your company has budget for such things, then check out https://asciinema.org/consulting

I hope it answers your question.

**Comment by amityahav** [comment]
So those snapshots can be obtained by the lines() function? Also will the lines contain only the text without the escapes and char codes?

**Comment by ku1ik** [comment]
The lines contain "segments", which are text fragments with their associated color/background attributes. Text of each segment is just String of letters, without escape chars or anything non-printable.

**Comment by amityahav** [comment]
so it baiscally outputs the entire grid divided by lines?

**Reply by ku1ik** [reply]
`vt.lines()` takes the grid, divides it into lines, then subsequent characters with the same color attributes are grouped together. In JSON it would look something like this:

```json
[
  [["hello", { "fg": 1 }], ["world", {}], ["!", { "bg": 2 }]],
  [["foo        ", {}]],
  [["bar", { "bold": true }], ["baz     ", {}]],
]
```

This represents terminal of size 11x3.

**Comment by ku1ik** [comment]
In v0.8.0, which I just pushed, you can get full text of the terminal, including scrollback, with new `vt.text() -> Vec<String>` method. And to get text of visible lines (wrapped, without scrollback) you can do this:

```rust
vt.lines().iter().map(|line| line.text()).collect::<Vec<_>>()
```

**Reply by ku1ik** [reply]
Update: in v0.8.3 to get text of visible lines (view) you do this:

```rust
vt.view().iter().map(|line| line.text()).collect::<Vec<_>>()
```
